### PR TITLE
Implement Create and Update Blood Donation Request with Validation and Testing

### LINKED ISSUE #174: Implement Create and Update Blood Donation Request
**Description**: 
As part of the core functionality for BloodConnect, the following tasks have been completed for the donation request generation workflow:
- Implemented both create and update functionality in the API Gateway.
- Developed a validation Lambda function for input data.
- Successfully saved the donation requests to DynamoDB for both creation and updates.

**Acceptance Criteria**:
- Successful creation of donation requests through the API.
- Validation of input data prior to saving to DynamoDB.
- Successful updating of existing donation requests.

### PR DESCRIPTION
**Pull request by EbrahimAli-RU**
Fixes #174 

**Donation Request Generation & Update Workflow**
- Added Lambda functions for creating and updating blood donation requests.
- Integrated validation logic, including geohash for location-based data.
- Created API Gateway endpoints with authorizers for secure access.
- Updated DynamoDB with GSI and LSI to support the new data model.
- Added unit tests for Lambda functions, service, and validation.
- 
- [x] Does not depend on other PRs
- [ ] Depends on other PR being merged first: #

## What kind of change does this PR introduce (check at least one)

- [ ] Bugfix
- [x] Improvement
- [ ] New feature
- [ ] Refactor
- [ ] Infrastructure
- [ ] Other (please specify)

## How would you describe the changes in a changelog (1-2 lines)?

Provide a brief overview of the changes introduced in this pull request.

## Related Issues/PRs (optional)

Link to the relevant issue(s) or provide a clear description of the problem or feature being addressed.

## How to test

- [ ] Needs regression testing (add QA needed label)

## Checklist

- [x] I have tested these changes locally and they function as expected.
- [ ] I have added or updated necessary documentation to reflect the changes.
- [x] I have followed the project's coding style and guidelines.
- [x] I have added appropriate test coverage to ensure the changes are functioning correctly.
- [x] I have checked for any potential conflicts with other branches or pull requests.
- [ ] I have updated necessary OpenAPI specs

## Reviewer Request (optional)

If you have any specific requests or suggestions for reviewers, mention them here.

## Screenshots (optional)

If applicable, provide any additional screenshots or GIFs that showcase the changes made in this pull request.

## Demo Link (optional)

If applicable, provide a link to a live demo or deployment where the changes can be reviewed.

## Known Issues (optional)

If there are any known issues or limitations with the changes made in this pull request, describe them here.

## Additional Notes (optional)

Add any additional notes or comments you think are relevant to this pull request.

## Deploy risk & follow up

Please assess risk & follow up on deployment of this PR. Elaborate your choice if selecting anything other than 'No risk'.

- [x] No risk
- [ ] Low
- [ ] High

### CODE REVIEW

File: core/application/bloodDonationWorkflow/BloodDonationService.ts [hunk 1]
```diff
@@ -0,0 +1,58 @@
+import { GENERIC_CODES } from '../../../commons/libs/constants/GenericCodes'
+import BloodDonationOperationError from './BloodDonationOperationError'
+import { DonationDTO } from '../../../commons/dto/DonationDTO'
+import { generateUniqueID } from '../utils/idGenerator'
+import Repository from '../technicalImpl/policies/repositories/Repository'
+import { generateGeohash } from '../utils/geohash'
+import { validateInputWithRules } from '../utils/validator'
+import { BloodDonationAttributes, validationRules, UpdateBloodDonationAttributes } from './Types'
+
+export class BloodDonationService {
```
Comment (sakibstark11) [inline core/application/bloodDonationWorkflow/BloodDonationService.ts:10]: why is this a class? it doesn't even have a constructor

File: core/application/bloodDonationWorkflow/BloodDonationService.ts [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+  async createBloodDonation(donationAttributes: BloodDonationAttributes, bloodDonationRepository: Repository<DonationDTO>): Promise<string> {
+    try {
+      const validationResponse = validateInputWithRules({ bloodQuantity: donationAttributes.bloodQuantity, donationDateTime: donationAttributes.donationDateTime }, validationRules)
+      if (validationResponse !== null) {
+        return validationResponse
+      }
+      await bloodDonationRepository.create({
+        id: generateUniqueID(),
+        ...donationAttributes,
+        status: 'accepted',
```
Comment (sakibstark11) [inline core/application/bloodDonationWorkflow/BloodDonationService.ts:20]: why are we using hardcoded values when you've declared an enum

File: core/application/bloodDonationWorkflow/BloodDonationService.ts [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+        geohash: generateGeohash(donationAttributes.latitude, donationAttributes.longitude),
+        donationDateTime: new Date(donationAttributes.donationDateTime).toISOString()
+      })
+      return 'We have accepted you request we will let you know when we will found donar.'
```
Comment (sakibstark11) [inline core/application/bloodDonationWorkflow/BloodDonationService.ts:24]: incorrect grammer. 
We have accepted you request and we will let you know when we find a donor.

maybe send back a request id as well

File: core/application/bloodDonationWorkflow/BloodDonationService.ts [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+    }
+  }
+
+  async updateBloodDonation(donationAttributes: UpdateBloodDonationAttributes, bloodDonationRepository: Repository<DonationDTO>): Promise<string> {
+    try {
+      const { requestPostId, donationDateTime, ...restAttributes } = donationAttributes
+
+      const item = await bloodDonationRepository.getItem(`BLOOD_REQ#${donationAttributes.seekerId}`, `BLOOD_REQ#${requestPostId}`)
+      if (item === null) {
+        return 'No item found.'
```
Comment (sakibstark11) [inline core/application/bloodDonationWorkflow/BloodDonationService.ts:36]: item not found

File: core/application/bloodDonationWorkflow/BloodDonationService.ts [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+      }
+      if (item?.status !== undefined && item.status === 'completed') {
+        return 'This request is completed you can\'t update'
```
Comment (sakibstark11) [inline core/application/bloodDonationWorkflow/BloodDonationService.ts:39]: incorrect grammar. 
You can't update a completed request

File: core/application/bloodDonationWorkflow/BloodDonationService.ts [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+      }
+      if (donationDateTime !== undefined) {
+        updateData.donationDateTime = new Date(donationDateTime).toISOString()
+        const validationResponse = validateInputWithRules({ donationDateTime }, validationRules)
+        if (validationResponse !== null) {
+          return validationResponse
+        }
+      }
+      await bloodDonationRepository.update(updateData)
+      return 'We have updated your request, we will let you know update.'
```
Comment (sakibstark11) [inline core/application/bloodDonationWorkflow/BloodDonationService.ts:53]: incorrect grammar.

File: core/application/tests/bloodDonationWorkflow/BloodDonationService.test.ts [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+      (generateGeohash as jest.Mock).mockReturnValue('geohash123');
+      (validateInputWithRules as jest.Mock).mockReturnValue(null)
+
+      const result = await bloodDonationService.createBloodDonation(donationAttributes, bloodDonationRepository)
+
+      expect(generateUniqueID).toHaveBeenCalled()
+      expect(generateGeohash).toHaveBeenCalledWith(donationAttributes.latitude, donationAttributes.longitude)
+      expect(bloodDonationRepository.create).toHaveBeenCalledWith(expect.objectContaining({
+        id: 'uniqueID',
+        status: 'accepted',
```
Comment (sakibstark11) [inline core/application/tests/bloodDonationWorkflow/BloodDonationService.test.ts:43]: use enum

**Comment by sakibstark11** [comment]
Please go through all the sentences you've used and correct their grammar. Please use enums where-ever you can. Please go through your code again.

File: core/application/bloodDonationWorkflow/BloodDonationService.ts (hunk 2 above)
Comment (EbrahimAli-RU) [reply core/application/bloodDonationWorkflow/BloodDonationService.ts]: Create an enum and use it in code. [Commit](https://github.com/CraftsmenLtd/BloodConnect/pull/177/commits/be92e1051e74df3cdc6edffb1aad6cfb725c40e8)

File: core/application/bloodDonationWorkflow/BloodDonationService.ts (hunk 4 above)
Comment (EbrahimAli-RU) [reply core/application/bloodDonationWorkflow/BloodDonationService.ts]: Updated to `Item not found.`

File: core/application/bloodDonationWorkflow/BloodDonationService.ts (hunk 5 above)
Comment (EbrahimAli-RU) [reply core/application/bloodDonationWorkflow/BloodDonationService.ts]: Updated to `You can't update a completed request`

File: core/application/bloodDonationWorkflow/BloodDonationService.ts (hunk 6 above)
Comment (EbrahimAli-RU) [reply core/application/bloodDonationWorkflow/BloodDonationService.ts]: Fix grammar. `We have updated your request and will let you know once there is an update.`

File: core/application/tests/bloodDonationWorkflow/BloodDonationService.test.ts (hunk 7 above)
Comment (EbrahimAli-RU) [reply core/application/tests/bloodDonationWorkflow/BloodDonationService.test.ts]: Create an enum and use it in code. [Commit](https://github.com/CraftsmenLtd/BloodConnect/pull/177/commits/be92e1051e74df3cdc6edffb1aad6cfb725c40e8)

File: core/application/bloodDonationWorkflow/BloodDonationService.ts (hunk 3 above)
Comment (EbrahimAli-RU) [reply core/application/bloodDonationWorkflow/BloodDonationService.ts]: Vai, This is my understanding, but I might be completely wrong.

After we create a blood request post and it succeeds, we might be directed to a page displaying all the posts. At that point, the new post we just created will also be visible, along with its request ID.

File: core/application/bloodDonationWorkflow/Types.ts [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+  location: string;
+  latitude: number;
+  longitude: number;
+  donationDateTime: string;
+  contactInfo: { name: string; phone: string };
+  patientName?: string;
+  transportationInfo?: string;
+  shortDescription?: string;
+}
+type CredentialKeys = 'donationDateTime' | 'bloodQuantity'
```
Comment (sakibstark11) [inline core/application/bloodDonationWorkflow/Types.ts:18]: why is this called credentialKeys?

File: core/application/tests/bloodDonationWorkflow/BloodDonationService.test.ts [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+describe('BloodDonationService', () => {
+  const bloodDonationService = new BloodDonationService()
+  const bloodDonationRepository = importedMockRepository as jest.Mocked<Repository<DonationDTO>>
+
+  beforeEach(() => {
+    jest.clearAllMocks()
+  })
+
+  describe('createBloodDonation', () => {
+    test('should create a blood donation and return success message', async() => {
```
Comment (sakibstark11) [inline core/application/tests/bloodDonationWorkflow/BloodDonationService.test.ts:32]: again grammar issue

**Review by sakibstark11 — APPROVED** [review]

**Review by KhirulIslam — APPROVED** [review]

**Pull request merged**